velox/exec/Exchange.cpp

```diff
@@ -226,22 +210,22 @@ void ExchangeClient::noMoreRemoteTasks() {
 }
 
 void ExchangeClient::close() {
-  std::vector<std::shared_ptr<ExchangeSource>> sources;
-
   {
-    std::lock_guard<std::mutex> l(queue_->mutex());
-
-    if (closed_) {
-      return;
+    std::vector<std::shared_ptr<ExchangeSource>> sources;
+    {
```

Comment:
This is not safe as the source could still add new page into queue_ after we release the lock and before we call source close. Consider
```
std::vector<std::shared_ptr<ExchangeSource>> sources;
{
   std::lock_guard<std::mutex> l(queue_->mutex());
   if (closed_) {
      return;
   }
   closed_ = true;
   sources = std::move(sources_);
}
// Close the sources outside of queue lock to avoid deadlock.
for (auto& source : sources) {
   source->close();
}
// Close the queue by discarding any buffered pages after we close all the sources.
{
   std::lock_guard<std::mutex> l(queue_->mutex());
   queue_->closeLock();
}
```